src/java.base/share/classes/java/util/concurrent/locks/AbstractQueuedSynchronizer.java

```diff
@@ -1553,6 +1610,25 @@ private void unlinkCancelledWaiters(ConditionNode node) {
             }
         }
 
+        /**
+         * Constructs objects needed for condition wait. On OOME,
+         * releases lock, sleeps, reacquires, and returns null.
+         */
+        private ConditionNode newConditionNode() {
+            int savedState;
+            if (tryInitializeHead() != null) {
+                try {
+                    return new ConditionNode();
+                } catch (OutOfMemoryError oome) {
+                }
+            }
+            if (!isHeldExclusively() || !release(savedState = getState()))
+                throw LockSupport.staticIllegalMonitorStateException; // OOM
```

Comment:
How is it possible to get IMSE this deep into the code? And the comment is confusing - OOM?

Reply:
Clarified to:
            // fall through if encountered OutOfMemoryError
            if (!isHeldExclusively() || !release(savedState = getState()))
                throw LockSupport.staticIllegalMonitorStateException;